[Reproduce tool] Add basic support for Android.

mbarbella-chromium opened the pull request.

mbarbella-chromium commented:
/gcbrun

mbarbella-chromium commented on src/local/butler/reproduce.py:
```diff
@@ -208,6 +209,28 @@ def _get_testcase_id_from_url(testcase_url):
+    # TODO(mbarbella): Handle the one-device case gracefully.
+    raise errors.ReproduceToolUnrecoverableError('Please set ANDROID_SERIAL.')
+
+  # TODO(mbarbella): Should we run device.initialize_device?
```
This will be trivial to add, but some of the settings changes and other modifications seem invasive. Should we do it anyway?

oliverchang replied:
Maybe add a warning/prompt and then do this?

inferno-chromium replied:
+1 to prompt. And then we need to call initialize_device, otherwise things wont work, like setup_host_and_device_forwarder_if_needed, setup_asan_if_needed are all called from there. either can create another lightweight version of the function for just reproduce itself, like which calls e.g. setup_adb, etc.

mbarbella-chromium replied:
Added a warning. Based on Abhishek's point I think it's fine to always require that this happen to continue, but the prompt is still a great idea to make sure users know what they're getting into.

mbarbella-chromium commented on src/python/bot/tasks/setup.py:
```diff
@@ -63,33 +63,35 @@ def _copy_testcase_to_device_and_setup_environment(testcase,
 
   # The following steps need privileged job access.
   job_type_has_privileged_access = environment.get_value('PRIVILEGED_ACCESS')
-  if job_type_has_privileged_access:
```
There should be no functional changes here, just noticed that it was formatted in a way that would have been better served by an early bailout.

mbarbella-chromium commented:
/gcbrun

mbarbella-chromium commented:
PTAL when you have time. I still want to test this a bit more, so no rush.

mbarbella-chromium commented:
/gcbrun

mbarbella-chromium commented:
/gcbrun

inferno-chromium commented:
@oliverchang - can you help to review.

oliverchang commented on src/local/butler/reproduce.py:
```diff
@@ -220,6 +243,16 @@ def _reproduce_crash(testcase_url, build_directory):
+  # Validate that we're running on the right platform for this test case.
+  platform = environment.platform().lower()
+  if testcase.platform == 'android' and platform == 'linux':
+    _prepare_environment_for_android(testcase_path)
```
maybe re-structure the conditions a little here such that we give a better error condition if testcase platform is 'android' but environment platform != 'linux'?

inferno-chromium commented on src/local/butler/reproduce.py:
```diff
@@ -208,6 +209,28 @@ def _get_testcase_id_from_url(testcase_url):
+  # TODO(mbarbella): Should we run device.initialize_device?
+
+  # Push the test case and build APK to the device.
+  apk_path = environment.get_value('APP_PATH')
```
add assert or something to check if APK_PATH is set correctly.

inferno-chromium commented on src/local/butler/reproduce.py:
```diff
@@ -208,6 +209,28 @@ def _get_testcase_id_from_url(testcase_url):
+
+  # Push the test case and build APK to the device.
+  apk_path = environment.get_value('APP_PATH')
+  device.install_application_if_needed(apk_path, True)
```
use named argument force_update=True

inferno-chromium commented on src/local/butler/reproduce.py:
```diff
@@ -208,6 +209,28 @@ def _get_testcase_id_from_url(testcase_url):
+
+  testcase_manager.get_command_line_for_application(
+      write_command_line_file=True)
+
```
PKG_NAME is no set, which will cause gestures to not work. 

https://github.com/google/clusterfuzz/blob/master/src/python/platforms/android/gestures.py#L39

inferno-chromium commented on src/local/butler/reproduce.py:
```diff
@@ -208,6 +209,28 @@ def _get_testcase_id_from_url(testcase_url):
+  testcase_manager.get_command_line_for_application(
+      write_command_line_file=True)
+
+  environment.set_value('FUZZ_INPUTS', os.path.dirname(testcase_path))
```
Why is this android specific, FUZZ_INPUTS should be set for all.

inferno-chromium commented on src/local/butler/reproduce.py:
```diff
@@ -208,6 +209,28 @@ def _get_testcase_id_from_url(testcase_url):
+
+  # TODO(mbarbella): Should we run device.initialize_device?
+
+  # Push the test case and build APK to the device.
```
Why not move all this code from
https://github.com/google/clusterfuzz/blob/442a074fefbe5fdcbbe0df046e985c850c64a73e/src/python/build_management/build_manager.py#L665 into android device.py and then use the helper here. we need to avoid code duplication in reproduce tool as much as possible as otherwise we will have more maintenance later.

mbarbella-chromium commented:
/gcbrun

mbarbella-chromium commented:
There's still one open comment thread regarding PKG_NAME, but otherwise I think I've addressed all comments. This still needs more local testing, but should be close to ready.

inferno-chromium commented:
> There's still one open comment thread regarding PKG_NAME, but otherwise I think I've addressed all comments. This still needs more local testing, but should be close to ready.

ah right, can ignore this one. PKG_NAME is always in job and other case https://github.com/google/clusterfuzz/blob/299f2e2ba20c4c6eb1596fe350f58c9f1bd52c14/src/python/bot/tasks/setup.py#L71 is custom apk and uncommon, can skip for now.

inferno-chromium commented on src/local/butler/reproduce.py:
```diff
@@ -176,6 +178,10 @@ def _update_environment_for_testcase(testcase, build_directory):
   app_path = os.path.join(build_directory, environment.get_value('APP_NAME'))
   environment.set_value('APP_PATH', app_path)
 
+  if not os.path.exists(app_path):
```
if not app_path or not os.path.exists

mbarbella-chromium replied:
Factored this out into a new function (for mocking) and used a default value instead.

inferno-chromium commented on src/local/butler/reproduce_tool/prompts.py:
```diff
@@ -0,0 +1,28 @@
+
+def get_boolean(prompt):
+  """Return a boolean representing a yes/no answer to a prompt."""
+  result = get_string(prompt + ' (Y/n): ', add_colon=False)
```
Why not just do ' (Y/n)' so that colon is added always ?

mbarbella-chromium replied:
Done

inferno-chromium commented on src/python/platforms/android/device.py:
```diff
@@ -438,3 +438,20 @@ def setup_host_and_device_forwarder_if_needed():
+  # On Android, we may need to write a command line file. We do this in
+  # advance so that we do not have to write this to the device multiple
+  # times.
+  # TODO(mbarbella): Build code should not depend on fuzzing.
```
Please remove this TODO, this is now done. also, can move "from fuzzing " into the module imports at top.

mbarbella-chromium replied:
Unfortunately we have the same problem in this file. testcase_manager depends on a lot of the rest of the codebase, so importing it from the files with these sorts of helper functions will always be problematic. Not sure what the best way to handle it is, but updated the TODO so it makes sense in this file.

mbarbella-chromium commented:
/gcbrun

mbarbella-chromium commented:
/gcbrun
